pkg/fleetautoscalers/fleetautoscalers.go

```diff
@@ -41,100 +41,134 @@ var client = http.Client{
 
 // computeDesiredFleetSize computes the new desired size of the given fleet
 func computeDesiredFleetSize(fas *autoscalingv1.FleetAutoscaler, f *agonesv1.Fleet) (int32, bool, error) {
-
 	switch fas.Spec.Policy.Type {
 	case autoscalingv1.BufferPolicyType:
 		return applyBufferPolicy(fas.Spec.Policy.Buffer, f)
 	case autoscalingv1.WebhookPolicyType:
 		return applyWebhookPolicy(fas.Spec.Policy.Webhook, f)
 	}
 
-	return f.Status.Replicas, false, errors.New("wrong policy type, should be one of: Buffer, Webhook")
+	return 0, false, errors.New("wrong policy type, should be one of: Buffer, Webhook")
 }
 
-func applyWebhookPolicy(w *autoscalingv1.WebhookPolicy, f *agonesv1.Fleet) (int32, bool, error) {
-	faReq := autoscalingv1.FleetAutoscaleReview{
-		Request: &autoscalingv1.FleetAutoscaleRequest{
-			UID:       uuid.NewUUID(),
-			Name:      f.Name,
-			Namespace: f.Namespace,
-			Status:    f.Status,
-		},
-		Response: nil,
+func applyWebhookPolicy(w *autoscalingv1.WebhookPolicy, f *agonesv1.Fleet) (replicas int32, limited bool, err error) {
+	if w == nil {
+		return 0, false, errors.New("nil WebhookPolicy passed")
 	}
-	b, err := json.Marshal(faReq)
-	urlStr := ""
-	if w.URL != nil {
-		urlStr = *w.URL
+
+	if f == nil {
+		return 0, false, errors.New("nil Fleet passed")
 	}
-	var faResp autoscalingv1.FleetAutoscaleReview
-	servicePath := ""
-	if w.Service != nil {
-		if w.Service.Path != nil {
-			servicePath = *w.Service.Path
+
+	if w.URL != nil && w.Service != nil {
+		return 0, false, errors.New("service and url cannot be used simultaneously")
+	}
+
+	var u *url.URL
```

Comment:
The section to set the url is complicated enough that it might be nice to split into a separate helper method (which could also make the control flow a bit simpler and reduce the number of levels of indentation).

Reply:
@roberthbailey is it better?

Reply:
Part of the reason I suggested making it a separate function is so that you could reduce the indentation and simplify the code flows (since you can return right away on errors).

```
func buildURLFromWebhookPolicy(w *autoscalingv1.WebhookPolicy) (*url.URL, error) {
  if w.URL != nil {
    if *w.URL == "" {
      return nil, errors.New("URL was not provided")
    }
    return url.ParseRequestURI(*w.URL)
  }

  // notice how the next block of code can now be at the top level since the previous logic has returned in all branches
  if w.Service.Name == "" {
    return nil, errors.New("service name was not provided")
  }
  var servicePath string
  if w.Service.Path != nil {
    servicePath = *w.Service.Path
  }
  if w.Service.Namespace == "" {
    w.Service.Namespace = "default"
  }
  scheme := "http"
  if w.CABundle != nil {
    scheme = "https"
    if err := setCABundle(w.CABundle); err != nil {
      return nil, err
    }
  }
  return &url.URL{
    Scheme: scheme,
    Host:   fmt.Sprintf("%s.%s.svc:8000", w.Service.Name, w.Service.Namespace),
    Path:   servicePath,
  }, nil
}
```

Reply:
Ok, now I get your point. Updated.